File: src/main/java/com/yauhenl/neuro/ai/Brain.java
package com.yauhenl.neuro.ai;

public class Brain {
    private final Perceptron[][] layers;
    private final Integer layersNum = 2;
    private final Integer perceptronNum = 4;

    public Brain(DNA dna) {
        layers = new Perceptron[layersNum][perceptronNum];
        for (int i = 0; i < layersNum; ++i) {
            for (int j = 0; j < perceptronNum; ++j) {
                float[] w = new float[4];
                w[0] = dna.getWeightGenes().get(16 * i + 4 * j);
                w[1] = dna.getWeightGenes().get(16 * i + 4 * j + 1);
                w[2] = dna.getWeightGenes().get(16 * i + 4 * j + 2);
                w[3] = dna.getWeightGenes().get(16 * i + 4 * j + 3);
                layers[i][j] = new Perceptron(perceptronNum, w);
            }
        }
    }

    public float[] computeOutput(float[] inputs) {
        float[][] out = new float[layersNum + 1][perceptronNum];
        out[0] = inputs;
        for (int i = 0; i < layersNum; ++i) {
            for (int j = 0; j < perceptronNum; ++j) {
                out[i + 1][j] = layers[i][j].compute(out[i]);
                out[i + 1][j] = layers[i][j].compute(out[i]);
                out[i + 1][j] = layers[i][j].compute(out[i]);
                out[i + 1][j] = layers[i][j].compute(out[i]);
            }
        }
        return out[layersNum];
    }
}

File: src/main/kotlin/com/yauhenl/kotlin/neuro/ai/Brain.kt
package com.yauhenl.kotlin.neuro.ai

class Brain(dna: DNA) {
    private val layers: Array<Array<Perceptron?>>
    private val layersNum = 2
    private val perceptronNum = 4

    init {
        layers = Array(layersNum, { arrayOfNulls<Perceptron>(perceptronNum)})
        for (i in 0..layersNum - 1) {
            for (j in 0..perceptronNum - 1) {
                val w = FloatArray(4)
                w[0] = dna.weightGenes[16 * i + 4 * j]
                w[1] = dna.weightGenes[16 * i + 4 * j + 1]
                w[2] = dna.weightGenes[16 * i + 4 * j + 2]
                w[3] = dna.weightGenes[16 * i + 4 * j + 3]
                layers[i][j] = Perceptron(perceptronNum, w)
            }
        }
    }

    fun computeOutput(inputs: FloatArray): FloatArray {
        val out = Array(layersNum + 1) { FloatArray(perceptronNum) }
        out[0] = inputs
        for (i in 0..layersNum - 1) {
            for (j in 0..perceptronNum - 1) {
                out[i + 1][j] = layers[i][j]!!.compute(out[i])
                out[i + 1][j] = layers[i][j]!!.compute(out[i])
                out[i + 1][j] = layers[i][j]!!.compute(out[i])
                out[i + 1][j] = layers[i][j]!!.compute(out[i])
            }
        }
        return out[layersNum]
    }
}
